go/store/datas/database_common.go

```diff
@@ -547,67 +555,192 @@ func (db *database) doUpdateWorkingSet(ctx context.Context, datasetID string, wo
 		return err
 	}
 
-	// This could loop forever, given enough simultaneous writers. BUG 2565
+	workingSetRef, err := db.WriteValue(ctx, workingSet) // will be orphaned if the tryCommitChunks() below fails
+	if err != nil {
+		return err
+	}
+
+	wsValRef, err := types.ToRefOfValue(workingSetRef, db.Format())
+	if err != nil {
+		return err
+	}
+
 	var tryCommitErr error
 	testSetFailed := false
 	for tryCommitErr = ErrOptimisticLockFailed; tryCommitErr == ErrOptimisticLockFailed && !testSetFailed; {
 		tryCommitErr = func() error {
 			db.mu.Lock()
 			defer db.mu.Unlock()
 
+			success, err := db.assertDatasetHash(ctx, datasetID, currHash)
+			if err != nil {
+				return err
+			}
+
+			if !success {
+				testSetFailed = true
+			}
+
+			currentDatasets, err := db.Datasets(ctx)
+			if err != nil {
+				return err
+			}
+
+			currentDatasets, err = currentDatasets.Edit().Set(types.String(datasetID), wsValRef).Map(ctx)
+			if err != nil {
+				return err
+			}
+
 			currentRootHash, err := db.rt.Root(ctx)
+			if err != nil {
+				return err
+			}
+
+			return db.tryCommitChunks(ctx, currentDatasets, currentRootHash)
+		}()
+	}
+
+	return tryCommitErr
+}
+
+// assertDatasetHash returns true if the hash of the dataset matches the one given. Use an empty hash for a dataset you
+// expect not to exist.
+// Typically this is called using pessimistic locking by the caller in order to implement atomic test-and-set semantics.
```

Comment:
```suggestion
// Typically this is called using optimistic locking by the caller in order to implement atomic test-and-set semantics.
```

Reply:
Going to make some changes as we discussed